### DISCUSSION TITLE
Level I Trades and Quotes (TAQ) Data

### DISCUSSION THREAD

**Discussion by BGTCapital** [bmoscon/cryptofeed#726] [Q&A]
Hi
We are trying to build a very simple dataframe in panda with basic bid-ask price and volume 
![Bid Ask Data](https://user-images.githubusercontent.com/94967120/144101584-62117073-0b18-4a26-b239-8cbedb4cd49c.png)

Could you kindly help us understand how can we do that with Cryptofeed? I am still struggling to understand the documentation, and since I am new to python, I would highly appreciate if you could share some sample code 
Thanks in advance for your support
Best

**Comment by bmoscon** [comment]
I'm not sure where this data would come from. Is this just the top level of the orderbook? If so, you can subscribe to an L2_BOOK channel and take the top level off the book in each callback and append it to a dataframe. The code in `examples/demo.py` does something like this because it uses the top level to calculate the spread (in the book callback).

**Reply by bmoscon** [reply]
the callback would be something like

```python
async def book(data, receipt_timestamp):
     ask, ask_size = data.book.asks.index(0)
     bid, bid_size = data.book.bids.index(0)
     df.append({'ask': ask, 'ask_size': ask_size, 'bid': bid, 'bid_size': bid_size, 'timestamp': data.timestamp if data.timestamp else receipt_timestamp}, inplace=True)
```

**Reply by BGTCapital** [reply]
Thank you very (very!) much for your support. Since I am new to pandas, I was wondering if there is any way to add the data to the dataframe. the code from above returns: TypeError: append() got an unexpected keyword argument 'inplace'
On the other side, If I use this definition:

async def book(data, receipt_timestamp):
     ask, ask_size = data.book.asks.index(0)
     bid, bid_size = data.book.bids.index(0)
     print({'ask': ask, 'ask_size': ask_size, 'bid': bid, 'bid_size': bid_size, 'timestamp': data.timestamp if data.timestamp else receipt_timestamp})

I receive this data

{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369936.971917}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.1687636}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.1692595}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.16953}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.1697989}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.1700351}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.170259}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.1704915}
{'ask': Decimal('58561'), 'ask_size': Decimal('0.00041392'), 'bid': Decimal('58547'), 'bid_size': Decimal('0.00695'), 'timestamp': 1638369937.1707687}


Sorry for asking again, but since I am new to this, I was wondering if you could be so kind to help me structuring the dataframe and saving to csv file so that multiple coins are aggregated as columns as in this example

[sample_file.csv](https://github.com/bmoscon/cryptofeed/files/7635115/sample_file.csv)
.
Thanks in advance for your support

**Reply by bmoscon** [reply]
ok yeah looks like append doesnt support inplace, so it would just be df = df.append(.....) (leaving out the inplace kawrg).
